File: ts/src/cursor.ts
// Cursor over Borsh-encoded bytes, shared by the account and event decoders.
// Walks the schema field by field; no hardcoded offsets.

import { HASH_LEN } from "./constants.js";

export class TruncatedError extends Error {
  constructor(readonly field: string) {
    super(`truncated while reading ${field}`);
    this.name = "TruncatedError";
  }
}

export class Cursor {
  private pos: number;

  constructor(
    private readonly data: Uint8Array,
    start = 0,
  ) {
    this.pos = start;
  }

  /** Bytes not yet consumed. */
  remaining(): number {
    return Math.max(this.data.length - this.pos, 0);
  }

  take(len: number, field: string): Uint8Array {
    if (this.pos + len > this.data.length) {
      throw new TruncatedError(field);
    }
    const slice = this.data.slice(this.pos, this.pos + len);
    this.pos += len;
    return slice;
  }

  u8(field: string): number {
    return this.take(1, field)[0]!;
  }

  u32Le(field: string): number {
    const b = this.take(4, field);
    return new DataView(b.buffer, b.byteOffset, 4).getUint32(0, true);
  }

  u64Le(field: string): bigint {
    const b = this.take(8, field);
    return new DataView(b.buffer, b.byteOffset, 8).getBigUint64(0, true);
  }

  array32(field: string): Uint8Array {
    return this.take(HASH_LEN, field);
  }

  /** Borsh `Vec<u8>`: u32 LE length, then the bytes. */
  vecU8(field: string): Uint8Array {
    return this.take(this.u32Le(field), field);
  }

  /** Borsh `Vec<[u8; 32]>`: u32 LE length, then the arrays. */
  vecArray32(field: string): Uint8Array[] {
    const len = this.u32Le(field);
    const out: Uint8Array[] = [];
    for (let i = 0; i < len; i++) {
      out.push(this.array32(field));
    }
    return out;
  }
}

File: rust/src/cursor.rs
//! Cursor over Borsh-encoded bytes, shared by the account and event decoders.
//! Walks the schema field by field; no hardcoded offsets.

use crate::constants::HASH_LEN;

/// The data ran out while reading the named field.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct Truncated {
    pub field: &'static str,
}

pub struct Cursor<'a> {
    data: &'a [u8],
    pos: usize,
}

impl<'a> Cursor<'a> {
    pub fn new(data: &'a [u8], pos: usize) -> Self {
        Self { data, pos }
    }

    /// Bytes not yet consumed.
    pub fn remaining(&self) -> usize {
        self.data.len().saturating_sub(self.pos)
    }

    pub fn take(&mut self, len: usize, field: &'static str) -> Result<&'a [u8], Truncated> {
        let end = self.pos.checked_add(len).ok_or(Truncated { field })?;
        let slice = self.data.get(self.pos..end).ok_or(Truncated { field })?;
        self.pos = end;
        Ok(slice)
    }

    pub fn u8(&mut self, field: &'static str) -> Result<u8, Truncated> {
        Ok(self.take(1, field)?[0])
    }

    pub fn u32_le(&mut self, field: &'static str) -> Result<u32, Truncated> {
        let bytes: [u8; 4] = self.take(4, field)?.try_into().expect("4 bytes");
        Ok(u32::from_le_bytes(bytes))
    }

    pub fn u64_le(&mut self, field: &'static str) -> Result<u64, Truncated> {
        let bytes: [u8; 8] = self.take(8, field)?.try_into().expect("8 bytes");
        Ok(u64::from_le_bytes(bytes))
    }

    pub fn array_32(&mut self, field: &'static str) -> Result<[u8; 32], Truncated> {
        Ok(self.take(HASH_LEN, field)?.try_into().expect("32 bytes"))
    }

    /// Borsh `Vec<u8>`: u32 LE length, then the bytes.
    pub fn vec_u8(&mut self, field: &'static str) -> Result<Vec<u8>, Truncated> {
        let len = self.u32_le(field)? as usize;
        Ok(self.take(len, field)?.to_vec())
    }

    /// Borsh `Vec<[u8; 32]>`: u32 LE length, then the arrays.
    pub fn vec_array_32(&mut self, field: &'static str) -> Result<Vec<[u8; 32]>, Truncated> {
        let len = self.u32_le(field)? as usize;
        (0..len).map(|_| self.array_32(field)).collect()
    }
}
